libparsec/src/invite.rs

```diff
@@ -974,6 +1142,53 @@ pub async fn greeter_device_in_progress_1_do_wait_peer_trust(
         })
     };
 
+    let (cancel_requested, _canceller_guard) = listen_canceller(canceller)?;
+    libparsec_platform_async::select2_biased!(
+        res = work => res,
+        _ = cancel_requested => {
+            if canceller_ctx.cancel().await.is_err(){
+                // TODO: Warn about the error before discarding it
+            }
+            Err(GreetInProgressError::Cancelled)
+        },
+    )
+}
+
+pub async fn greeter_user_in_progress_2_do_deny_trust(
+    canceller: Handle,
+    handle: Handle,
+) -> Result<(), GreetInProgressError> {
+    let work = async {
+        let ctx = take_and_close_handle(handle, |x| match x {
+            HandleItem::UserGreetInProgress2(ctx) => Ok(ctx),
+            invalid => Err(invalid),
+        })?;
+
+        ctx.do_deny_trust().await?;
+        Ok(())
+    };
+
+    let (cancel_requested, _canceller_guard) = listen_canceller(canceller)?;
+    libparsec_platform_async::select2_biased!(
+        res = work => res,
+        _ = cancel_requested => Err(GreetInProgressError::Cancelled),
+    )
+}
+
+pub async fn greeter_device_in_progress_2_do_deny_trust(
+    canceller: Handle,
+    handle: Handle,
+) -> Result<(), GreetInProgressError> {
+    let work = async {
+        let ctx = take_and_close_handle(handle, |x| match x {
+            HandleItem::DeviceGreetInProgress2(ctx) => Ok(ctx),
+            invalid => Err(invalid),
+        })?;
+
+        ctx.do_deny_trust().await?;
+        Ok(())
+    };
```

Comment:
It might be possible using a macro, but note that all functions do not do exactly the same thing. Some return the next `Info` type, some don't and some have canceller, and some don't.

Plus, it's outside of the scope of this PR.